### PR TITLE
Specify the textInput event

### LINKED ISSUE #353: Specify the `textInput` event
Chromium and WebKit support a `textInput` event, which is similar to `beforeinput`. Firefox doesn't support it (yet), but this is a common web compat issue, so we plan to implement it.

Bugs:
- https://bugzilla.mozilla.org/show_bug.cgi?id=903746
- https://bugs.chromium.org/p/chromium/issues/detail?id=701480
- https://bugs.webkit.org/buglist.cgi?bug_status=UNCONFIRMED&bug_status=NEW&bug_status=ASSIGNED&bug_status=REOPENED&bug_status=RESOLVED&bug_status=VERIFIED&bug_status=CLOSED&list_id=9349913&product=WebKit&query_format=advanced&short_desc=textinput%20event&short_desc_type=allwordssubstr

IDL:
- https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/events/text_event.idl
- https://github.com/WebKit/WebKit/blob/main/Source/WebCore/dom/TextEvent.idl

One interop difference I've found so far is the order of events. Demo: https://software.hixie.ch/utilities/js/live-dom-viewer/saved/11385

- Chrome: beforeinput is fired first, then textInput.
-  Safari: textInput is fired first, then beforeinput.

### PR DESCRIPTION
**Pull request by zcorpan**
Fixes #353

The following tasks have been completed:

 * [x] Confirmed there are no ReSpec/BikeShed errors or warnings.
 * [x] Modified Web platform tests https://github.com/web-platform-tests/wpt/pull/44472

Implementation commitment:

 * [x] WebKit (already supported; event order: https://bugs.webkit.org/show_bug.cgi?id=268988)
 * [x] Chromium (already supported; minor IDL difference: https://issues.chromium.org/u/1/issues/324276564)
 * [x] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=903746)

### CODE REVIEW

File: sections/legacy-event-types.txt [hunk 1]
```diff
@@ -517,6 +518,118 @@ completeness.
...
+<h3 id="legacy-textevent-events">Legacy {{TextEvent}} events</h3>
+
+<h4 id="interface-textevent">Interface TextEvent</h4>
+
+
+	<pre class="idl">
+	[Exposed=Window]
+	interface TextEvent : Event {
+		readonly attribute DOMString data;
+		void initTextEvent(DOMString type,
```
Comment (zcorpan) [inline sections/legacy-event-types.txt:530]:
```suggestion
		undefined initTextEvent(DOMString type,
```

**Comment by zcorpan** [comment]
I think fixing https://github.com/w3c/uievents/issues/361 first would be nice so that this event can slot in between `beforeinput` and `input` in the default action chain.

File: event-algo.bs [hunk 2]
```diff
@@ -2073,10 +2183,16 @@ Issue: Should these be moved into the Clipboard spec?
...
+        Note: The "textInput" event is obsolete.
+
+    1. If |result| is false, then return.
 
-       1. Paste clipboard contents into DOM target element
-       1. <a>Fire an InputEvent</a> with "input", "insertFromPaste" and |data|
+    1. Paste clipboard contents into DOM target element
+    1. <a>Fire an InputEvent</a> with "input", "insertFromPaste" and |data|
 
 </div><!-- algorithm -->
```
Comment (masayuki-nakano) [inline event-algo.bs:2197]: If `inputType` of `beforeinput` is `insertLineBreak`, `textInput` is fired with `data="\n"` on Chrome.

Comment (masayuki-nakano) [reply event-algo.bs:2197]: If `inputType` of `beforeinput` is `insertCompositionText` **and** the composition is being committed, `textInput` is fired with `data=beforeinput.data`. On Chrome, `isComposing` of `beforeinput` and `input` are `true`. Unfortunately, I think that there is no way to define this behavior with attributes of the event. See also <https://https://github.com/w3c/uievents/issues/202>.

File: event-algo.bs [hunk 3]
```diff
@@ -2073,10 +2183,16 @@ Issue: Should these be moved into the Clipboard spec?
     1. Let |data| = the text on the clipboard being pasted
     1. Let |result| = <a>fire an InputEvent</a> with "beforeinput", "insertFromPaste" and |data|
```
Comment (masayuki-nakano) [inline event-algo.bs:2184]: How about `insertFromPasteAsQuotation`? It's currently supported only by Gecko.

File: event-algo.bs (hunk 2 above)
Comment (masayuki-nakano) [reply event-algo.bs:2197]: Odd, in the case of DnD, Chrome dispatches `textInput` immediately after `drop`, but before `beforeinput` whose `inputType` is (`deleteByDrag` and) `insertFromDrop`. Additionally, the event target of `textInput` is not the editing host.

I think that Gecko cannot emulate this behavior.

**Review by masayuki-nakano — CHANGES_REQUESTED** [review]

**Comment by masayuki-nakano** [comment]
And I've not tested when replacing typo word with the spellchecker.

File: event-algo.bs (hunk 2 above)
Comment (zcorpan) [reply event-algo.bs:2205]: `textInput` when composition is committed: 2a3cbd8

Comment (zcorpan) [reply event-algo.bs:2207]: `insertLineBreak` and `insertParagraph` cases fixed in f36ff37

Comment (masayuki-nakano) [reply event-algo.bs:2207]: (I'm still waiting the DnD definition and adding `insertFromPasteAsQuotation` after `insertFromPaste` (I mean `s/"insertFromPaste"/"insertFromPaste" or "insertFromPasteAsQuotation"/`)

File: sections/legacy-event-types.txt [hunk 4]
```diff
@@ -517,6 +518,118 @@ completeness.
...
 		</div>
 
+<h3 id="legacy-textevent-events">Legacy {{TextEvent}} events</h3>
+
+<h4 id="interface-textevent">Interface TextEvent</h4>
+
+
+	<pre class="idl">
+	[Exposed=Window]
+	interface TextEvent : Event {
```
Comment (masayuki-nakano) [inline sections/legacy-event-types.txt]: `TextEvent` inherits `UIEvent` in Chromium
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/events/text_event.h;l=38;drc=052831f0220b79fe0c3343b49f6d2863ea6de05d
and WebKit:
https://searchfox.org/wubkat/rev/77fb3649ec46dea08154a9f96e0d06fd5b9cb0bf/Source/WebCore/dom/TextEvent.idl#28

Comment (masayuki-nakano) [reply sections/legacy-event-types.txt]: FYI: `document.createEvent("TextEvent") instanceof UIEvent` returns `true` on Chrome.

**Comment by masayuki-nakano** [comment]
WPt is required at least for:
* no constructor
* `initTextEvent`
* `Document.createEvent` and `instance of TextEditor` and `instance of UIEvent` with the result.
* `addEventListener` with `textInput` and `textinput`
* With typing a character (an ASCII character and a non-BMP character)
* With deleting a character around caret (no `textInput`)
* With deleting a character with non-collapsed selection (no `textInput`)
* With `execCommand("insertText", false, "something")` (no `textInput`)

File: sections/legacy-event-types.txt (hunk 4 above)
Comment (zcorpan) [reply sections/legacy-event-types.txt]: Yep, I had noticed and fixed this already. :)
https://github.com/w3c/uievents/blob/f36ff37adac5bbbfa7bcd8b1b097dad5479cef2f/event-algo.bs#L958

File: event-algo.bs (hunk 2 above)
Comment (zcorpan) [reply event-algo.bs:2207]: Filed https://github.com/w3c/uievents/issues/367 for DnD.

File: event-algo.bs (hunk 3 above)
Comment (zcorpan) [reply event-algo.bs:2194]: I think that's a separate issue, filed https://github.com/w3c/uievents/issues/368

**Comment by masayuki-nakano** [comment]
Oh? There is no `ontextinput` nor `ontextInput` attribute in the global event handler? Chrome does not run the script set to it and there is no property when I check an element with the devtools.

**Comment by masayuki-nakano** [comment]
Hmm, in `<input>`, Chrome fires `beforeinput` event but no `textInput` event when I press `Enter`...

**Comment by masayuki-nakano** [comment]
When I replace a misspell word on Chrome, Chrome does not fire `textInput` after `beforeinput`.

**Comment by masayuki-nakano** [comment]
I wonder, whether `textInput` should be fired after a `beforeinput` event listener makes the editing host blurred.

File: event-algo.bs [hunk 5]
```diff
@@ -935,6 +940,109 @@ Issue: TODO
...
+    :: |eventTarget|, the {{EventTarget}} of the event
+
+    : Output
+    :: None
+
+    1. <a>Initialize a TextEvent</a> with |e|, |eventType| and |eventTarget|
+
+    1. Initialize the following public attributes
+
+        1. Set event.data = null
```
Comment (masayuki-nakano) [inline event-algo.bs:1008]: It's `DOMString`, but here sets `null`.

**Comment by masayuki-nakano** [comment]
If I call `preventDefault` of `textInput` at DnD in same editing host, the deletion is also canceled on Chrome. It's impossible to emulate that on Firefox without adding a lot of hacky code...

And also a call of `preventDefault` for IME composition cannot cancel the default action but its `cancelable` is set to `true` on Chrome.

**Comment by masayuki-nakano** [comment]
I wonder, how about `insertFromYank`? I don't know whether Chrome and/or Safari support it and how to do that.

**Comment by zcorpan** [comment]
> I wonder, whether `textInput` should be fired after a `beforeinput` event listener makes the editing host blurred.

Interestingly, Chrome and Safari seem to prevent moving focus while handling `beforeinput`: http://software.hixie.ch/utilities/js/live-dom-viewer/saved/12339

(Opening a popup works, but still fires the `textInput` event.)

**Comment by zcorpan** [comment]
> I wonder, how about `insertFromYank`? I don't know whether Chrome and/or Safari support it and how to do that.

I didn't know it was a thing, but I see it in https://w3c.github.io/input-events/ (issue: https://github.com/w3c/input-events/issues/20 ).

Per https://www.howtogeek.com/293850/how-to-use-macos-built-in-kill-and-yank-as-an-alternative-cut-and-paste/ on macOS you can use ctrl+k to kill (cut) and ctrl+y to yank (paste).

Testing in http://software.hixie.ch/utilities/js/live-dom-viewer/saved/12340 , neither Safari or Chrome fire `textInput` when yanking.

**Comment by zcorpan** [comment]
> Hmm, in `<input>`, Chrome fires `beforeinput` event but no `textInput` event when I press `Enter`...

Should I check if the event target is an `HTMLInputElement` element in the spec?

**Comment by masayuki-nakano** [comment]
> > Hmm, in `<input>`, Chrome fires `beforeinput` event but no `textInput` event when I press `Enter`...
> 
> Should I check if the event target is an `HTMLInputElement` element in the spec?

Yeah, I think so. (Thank you for investigating about yanking!)

**Comment by zcorpan** [comment]
> WPt is required at least for:
> 
>     * no constructor
> 
>     * `initTextEvent`
> 
>     * `Document.createEvent` and `instance of TextEditor` and `instance of UIEvent` with the result.
> 
>     * `addEventListener` with `textInput` and `textinput`
> 
>     * With typing a character (an ASCII character and a non-BMP character)
> 
>     * With deleting a character around caret (no `textInput`)
> 
>     * With deleting a character with non-collapsed selection (no `textInput`)
> 
>     * With `execCommand("insertText", false, "something")` (no `textInput`)

Thanks, tested in https://github.com/web-platform-tests/wpt/pull/44472

I used a smiley for non-BMP character, but I think it's not automatable with WebDriver currently.

**Comment by zcorpan** [comment]
> Should I check if the event target is an `HTMLInputElement` element in the spec?

Done in 7b9b6a5

**Comment by zcorpan** [comment]
> Oh? There is no `ontextinput` nor `ontextInput` attribute in the global event handler? Chrome does not run the script set to it and there is no property when I check an element with the devtools.

Indeed.

I think I've addressed all comments now.

**Review by garykac — APPROVED** [review]
Overall, this looks great. It's nice to see proper algorithms being added/improved.

**Pull request merged**